plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -227,29 +226,112 @@ func (g *genericScheduler) Preempt(pod *v1.Pod, nodeLister algorithm.NodeLister,
+			// Pass nodeToVictims to extender to do extra preemption logics. We can expect it returns a nodeToVictimsAfterExtender struct with:
+			//   1. Subset of the given nodes after Filter.
+			//   2. Additional victims on the nodes after preemption.
+			if nodeToVictimsAfterExtender, err := extender.ProcessPreemption(pod, nodeToVictims, g.cachedNodeInfoMap, pdbs); err != nil {
+				// If verb is not defined, fall back to call nodePassesExtendersForPreemption instead.
+				if strings.Contains(err.Error(), statusMethodNotAllowed) {
+					extenderPreemptionFailed = true
+					glog.Warningf("Extender preemption is not supported, fall back to check if candidate node is allowed by extender.")
+					break
+				} else {
+					return nil, nil, nil, err
+				}
+
+			} else {
+				// Write back result after extender to nodeToVictims.
+				for nodeName := range nodeToVictims {
+					// If node exists in nodeToVictims but does not exist in nodeToVictimsAfterExtender, delete it.
+					if victimAfterExtender, ok := nodeToVictimsAfterExtender[nodeName]; !ok {
+						delete(nodeToVictims, nodeName)
+					} else {
+						nodeToVictims[nodeName] = victimAfterExtender
+					}
+				}
+			}
+		}
+
+		if !extenderPreemptionFailed {
+			nodeName := pickOneNodeForPreemption(nodeToVictims)
+			if len(nodeName) == 0 {
+				return nil, nil, nil, err
+			}
+
+			// Lower priority pods nominated to run on this node, may no longer fit on
+			// this node. So, we should remove their nomination. Removing their
+			// nomination updates these pods and moves them to the active queue. It
+			// lets scheduler find another place for them.
+			nominatedPods := g.getLowerPriorityNominatedPods(pod, nodeName)
+
+			return g.cachedNodeInfoMap[nodeName].Node(), nodeToVictims[nodeName].Pods, nominatedPods, err
+		}
+	}
+
+	// Extender preemption failed, fall back to check if candidate node is allowed by extender.
 	for len(nodeToVictims) > 0 {
-		node := pickOneNodeForPreemption(nodeToVictims)
-		if node == nil {
+		nodeName := pickOneNodeForPreemption(nodeToVictims)
+		if len(nodeName) == 0 {
 			return nil, nil, nil, err
 		}
-		passes, pErr := nodePassesExtendersForPreemption(pod, node.Name, nodeToVictims[node].pods, g.cachedNodeInfoMap, g.extenders)
+		passes, pErr := nodePassesExtendersForPreemption(pod, nodeName, nodeToVictims[nodeName].Pods, g.cachedNodeInfoMap, g.extenders)
 		if passes && pErr == nil {
 			// Lower priority pods nominated to run on this node, may no longer fit on
 			// this node. So, we should remove their nomination. Removing their
 			// nomination updates these pods and moves them to the active queue. It
 			// lets scheduler find another place for them.
-			nominatedPods := g.getLowerPriorityNominatedPods(pod, node.Name)
-			return node, nodeToVictims[node].pods, nominatedPods, err
+			nominatedPods := g.getLowerPriorityNominatedPods(pod, nodeName)
+			return g.cachedNodeInfoMap[nodeName].Node(), nodeToVictims[nodeName].Pods, nominatedPods, err
```

Comment:
We should check that `g.cachedNodeInfoMap[nodeName]` is ok. The node could be deleted while scheduler is in the middle of the preemption work.